contracts/libs/data-structures/AvlTree.sol

```diff
@@ -0,0 +1,823 @@
+
+    function _removeMin(
+        mapping(bytes32 => Node) storage _tree,
+        bytes32 node_
+    ) private returns (bytes32) {
+        Node storage _node = _tree[node_];
+
+        if (_node.left == 0) {
+            return _node.right;
+        }
+
+        _node.left = _removeMin(_tree, _node.left);
+
+        return _balance(_tree, node_);
+    }
+
+    function _rotateLeft(
+        mapping(bytes32 => Node) storage _tree,
+        bytes32 node_
+    ) private returns (bytes32) {
+        Node storage _node = _tree[node_];
+
+        bytes32 temp_ = _node.left;
+        _node.left = _tree[temp_].right;
+        _tree[temp_].right = node_;
+
+        _updateHeight(_tree, node_);
+        _updateHeight(_tree, temp_);
+
+        return temp_;
+    }
+
+    function _rotateRight(
+        mapping(bytes32 => Node) storage _tree,
+        bytes32 node_
+    ) private returns (bytes32) {
+        Node storage _node = _tree[node_];
+
+        bytes32 temp_ = _node.right;
+        _node.right = _tree[temp_].left;
+        _tree[temp_].left = node_;
+
+        _updateHeight(_tree, node_);
+        _updateHeight(_tree, temp_);
+
+        return temp_;
+    }
+
+    function _balance(
+        mapping(bytes32 => Node) storage _tree,
+        bytes32 node_
+    ) private returns (bytes32) {
+        _updateHeight(_tree, node_);
+
+        Node storage _left = _tree[_tree[node_].left];
+        Node storage _right = _tree[_tree[node_].right];
+
+        if (_left.height > _right.height + 1) {
+            if (_tree[_left.right].height > _tree[_left.left].height) {
+                _tree[node_].left = _rotateRight(_tree, _tree[node_].left);
+            }
+
+            return _rotateLeft(_tree, node_);
+        } else if (_right.height > _left.height + 1) {
+            if (_tree[_right.left].height > _tree[_right.right].height) {
+                _tree[node_].right = _rotateLeft(_tree, _tree[node_].right);
+            }
+
+            return _rotateRight(_tree, node_);
+        }
+
+        return node_;
+    }
+
+    function _updateHeight(mapping(bytes32 => Node) storage _tree, bytes32 node_) private {
+        Node storage _node = _tree[node_];
+
+        _node.height = 1 + Math.max(_tree[_node.left].height, _tree[_node.right].height);
+    }
+
+    function _search(Tree storage tree, bytes32 key_) private view returns (bool) {
+        return key_ != 0 && tree.tree[key_].key == key_;
+    }
+
+    function _getMin(
+        mapping(bytes32 => Node) storage _tree,
+        bytes32 node_
+    ) private view returns (bytes32) {
+        while (_tree[node_].left != 0) {
+            node_ = _tree[node_].left;
+        }
+
+        return node_;
+    }
+
+    function _getMax(
+        mapping(bytes32 => Node) storage _tree,
+        bytes32 node_
+    ) private view returns (bytes32) {
+        while (_tree[node_].right != 0) {
+            node_ = _tree[node_].right;
+        }
+
+        return node_;
+    }
+
+    function _getTraversal(
```

Comment:
Using the `iterator` pattern will delegate the traversal responsibility to the special object (struct in our case). We will then implement a "traversal" library for these objects.